File: java-union-reader-writer-example/src/main/java/com/intellicess/union/java_union_reader_writer_example/Reader.java
package com.intellicess.union.java_union_reader_writer_example;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.util.DefaultIndenter;
import com.fasterxml.jackson.core.util.DefaultPrettyPrinter;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.intellicess.union.java_union_reader_writer_example.client.UnionClient;
import com.intellicess.union.java_union_reader_writer_example.client.UnionClientConfiguration;
import com.intellicess.union.java_union_reader_writer_example.client.model.JwlfResponse;
import com.intellicess.union.java_union_reader_writer_example.client.model.JwlfResponseList;
import com.intellicess.union.java_union_reader_writer_example.config.DevAppConfiguration;
import com.intellicess.union.java_union_reader_writer_example.config.EnvAppConfiguration;
import com.intellicess.union.java_union_reader_writer_example.utils.*;
import org.apache.logging.log4j.LogManager;
import org.apache.logging.log4j.Logger;

import java.io.File;
import java.io.IOException;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.StandardOpenOption;
import java.time.LocalDateTime;
import java.time.ZoneOffset;
import java.util.List;
import java.util.Optional;

public class Reader {

    private static final Logger log = LogManager.getLogger(Reader.class);

    public static void main(String[] args) throws InterruptedException, IOException {
        AppConfiguration config;
        try {
            config = EnvAppConfiguration.init();
        } catch (Exception e) {
            log.warn("Using reader dev configuration because of the following failure in initialization of" +
                    " environment variables based configuration:", e);
            config = DevAppConfiguration.initReader();
        }

        String readerLocalWorkingDirectory = config.getLocalWorkingDirectory();
        File readerDirectory = new File(readerLocalWorkingDirectory);
        if (!readerDirectory.isDirectory()) {
            throw new RuntimeException("com.intellicess.union.java_union_reader_writer_example.Reader local working directory with path='" + readerLocalWorkingDirectory + "' is not directory");
        }

        ObjectMapper objectMapper = ObjectMapperFactory.createObjectMapper();
        UnionClient unionClient = new UnionClient(
                UnionClientConfiguration.builder()
                        .unionUrl(config.getUnionUrl())
                        .authProviderUrl(config.getAuthProviderUrl())
                        .authProviderRealm(config.getAuthProviderRealm())
                        .authClientId(config.getAuthClientId())
                        .username(config.getAuthUserUsername())
                        .password(config.getAuthUserPassword())
                        .build(),
                objectMapper
        );

        String client = config.getUnionClient();
        String region = config.getUnionRegion();
        String asset = config.getUnionAsset();
        String folder = config.getUnionFolder();

        // if the following variable is set to null, it will read all data that was saved so far in given data space (i.e. in client/region/asset/folder)
        Long inclusiveSinceTimestamp = TimeUtils.dateTimeToMillisecondsTimestamp(LocalDateTime.now(ZoneOffset.UTC)) - 5000L;
        log.info("Listening to '{}/{}/{}/{}' data space in Union since timestamp='{}'...",
                client, region, asset, folder, inclusiveSinceTimestamp);
        while (!Thread.interrupted()) {
            JwlfResponseList responseList = unionClient.getNewJwlfLogsWithData(client, region, asset, folder, inclusiveSinceTimestamp);
            List<JwlfResponse> jwlfLogs = responseList.getList();
            for (JwlfResponse jwlfLog : jwlfLogs) {
                String filename = (String) jwlfLog.getHeader().getMetadata().get(CsvJwlfConverter.FILENAME_METADATA_KEY);

                boolean base64EncodedBinariesExample = Optional.ofNullable(jwlfLog
                        .getHeader()
                        .getMetadata()
                        .get(CsvJwlfWithBase64EncodedBinariesConverter.BASE64_ENCODED_BINARIES_EXAMPLE_METADATA_KEY))
                        .map(o -> Boolean.parseBoolean(o.toString()))
                        .orElse(false);
                if (base64EncodedBinariesExample) {
                    CsvJwlfWithBase64EncodedBinariesConverter.convertJwlfToFolder(readerLocalWorkingDirectory, jwlfLog);
                }

                if (!filename.endsWith(".json")) {
                    filename += ".json";
                }
                Path filePath = Path.of(readerLocalWorkingDirectory, filename);
                String json = toPrettyJson(objectMapper, jwlfLog);
                Files.writeString(filePath, json, StandardOpenOption.CREATE);
                log.info("Pulled '{}' and saved in local reader folder", filename);
            }
            Thread.sleep(1000L);
        }
    }

    private static String toPrettyJson(ObjectMapper objectMapper, JwlfResponse jwlfLog) throws JsonProcessingException {
        DefaultPrettyPrinter.Indenter indenter = new DefaultIndenter("    ", DefaultIndenter.SYS_LF);
        DefaultPrettyPrinter printer = new DefaultPrettyPrinter();
        printer.indentObjectsWith(indenter);
        printer.indentArraysWith(indenter);
        return objectMapper.writer(printer).writeValueAsString(jwlfLog);
    }
}

File: csharp-union-reader-writer-example/csharp-union-reader-writer-example/Reader.cs
using Serilog.Core;
using System.Net.Http.Json;
using System.Net.Http;
using System.Net.Http.Headers;
using System.Text;
using System.Text.Json;
using System.Text.Json.Serialization;
using Newtonsoft.Json;
using Union;

public class Reader 
{

    private static Logger log = AppLogFactory.CreateLogger();


    public static void Main() 
    {
        IAppConfiguration config;
        try
        {
            config = EnvAppConfiguration.Init();
        }
        catch(Exception e)
        {
            log.Warning(e, "Using reader dev configuration because of the following failure in initialization of environment variables based configuration:");
            config = DevAppConfiguration.InitReader();
        }
        string readerLocalWorkingDirectory = config.GetLocalWorkingDirectory();
        var readerDirectory = new DirectoryInfo(readerLocalWorkingDirectory);
        if(!readerDirectory.Exists)
        {
            throw new Exception("Reader local working directory with path='" + readerLocalWorkingDirectory + "' is not directory");
        }

        UnionClient unionClient = new UnionClient(
            new UnionClientConfiguration
            {
                UnionUrl=config.GetUnionUrl(),
                AuthProviderUrl=config.GetAuthProviderUrl(),
                AuthProviderRealm=config.GetAuthProviderRealm(),
                AuthClientId=config.GetAuthClientId(),
                Username=config.GetAuthUserUsername(),
                Password=config.GetAuthUserPassword()
            }
        );

        string client = config.GetUnionClient();
        string region = config.GetUnionRegion();
        string asset = config.GetUnionAsset();
        string folder = config.GetUnionFolder();

        // if the following variable is set to null, it will read all data that was saved so far in given data space (i.e. in client/region/asset/folder)
        long inclusiveSinceTimestamp = (DateTime.Now.Ticks / TimeSpan.TicksPerMillisecond) - 5000;
        
        log.Information("Listening to '{client}/{region}/{asset}/{folder}' data space in Union since timestamp='{inclusiveSinceTimestamp}'...",
            client, region, asset, folder, inclusiveSinceTimestamp);
        foreach(var jwlfStreamEvent in unionClient.GetJwlfsStream(client, region, asset, folder, inclusiveSinceTimestamp))
        {
            JwlfResponse jwlfLog = jwlfStreamEvent.Data.Content;
            string filename = jwlfLog.Header.Metadata["filename"];
            string base64EncodedBinariesKey = CsvJwlfWithBase64EncodedBinariesConverter.BASE64_ENCODED_BINARIES_EXAMPLE_METADATA_KEY;
            bool base64EncodedBinariesExample = jwlfLog.Header.Metadata.ContainsKey(base64EncodedBinariesKey)
                && bool.Parse(jwlfLog.Header.Metadata[base64EncodedBinariesKey]);
            if (base64EncodedBinariesExample) {
                CsvJwlfWithBase64EncodedBinariesConverter.ConvertJwlfToFolder(readerLocalWorkingDirectory, jwlfLog);
            }

            if (!filename.EndsWith(".json")) {
                filename += ".json";
            }
            string filepath = readerLocalWorkingDirectory + "/" + filename;
            SaveAsPrettyJson(jwlfLog, filepath);
            log.Information("Pulled '{filename}' and saved in local reader folder", filename);
        }
    }

    private static void SaveAsPrettyJson(JwlfResponse jwlfResponse, string filepath)
    {
        using (var file = File.CreateText(filepath))
        {
            string json = JsonConvert.SerializeObject(jwlfResponse, Formatting.Indented);
            file.Write(json);   
        }
    }
}
